File: js-port/fs-js/src/funcs/text/sub-string-function.js
const { BaseFunction, CallType } = require('../../core/function-base');
const helpers = require('../helpers');
const { FSDataType } = require('../../core/fstypes');

class SubStringFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'substring';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 3;
  }

  evaluate(provider, parameters) {
    if (parameters.count === 0) {
      return helpers.makeError(helpers.FsError.ERROR_PARAMETER_COUNT_MISMATCH, `${this.symbol} requires at least one parameter`);
    }

    const textResult = helpers.requireString(this.symbol, parameters.getParameter(provider, 0), 'string');
    if (!textResult.ok) {
      return textResult.error;
    }
    const text = textResult.value;

    let start = 0;
    if (parameters.count > 1) {
      const startResult = helpers.requireInteger(this.symbol, parameters.getParameter(provider, 1), 'index');
      if (!startResult.ok) {
        return startResult.error;
      }
      start = startResult.value;
    }

    let length = text.length - start;
    if (parameters.count > 2) {
      const lengthResult = helpers.requireInteger(this.symbol, parameters.getParameter(provider, 2), 'count');
      if (!lengthResult.ok) {
        return lengthResult.error;
      }
      length = lengthResult.value;
    }

    if (start < 0 || start >= text.length || length <= 0) {
      return helpers.makeValue(FSDataType.String, '');
    }

    const end = Math.min(text.length, start + length);
    return helpers.makeValue(FSDataType.String, text.substring(start, end));
  }
}

module.exports = {
  SubStringFunction
};

File: FunscScript/Functions/Text/SubStringFunction.cs
using FunscScript.Core;
using FunscScript.Model;

namespace FunscScript.Functions.Text
{
    public class SubStringFunction : IFsFunction
    {
        public int MaxParsCount => 3;
        public CallType CallType => CallType.Prefix;
        public string Symbol => "substring";
        public int Precedence => 0;

        public object Evaluate(IFsDataProvider parent, IParameterList pars)
        {
            if (pars.Count == 0)
                throw new Error.EvaluationTimeException($"{this.Symbol} requires at least one parameter.");

            var par0 = pars.GetParameter(parent, 0);
            var par1 = pars.Count > 1 ? pars.GetParameter(parent, 1) : null;
            var par2 = pars.Count > 2 ? pars.GetParameter(parent, 2) : null;

            var str = par0 as string;
            if (str == null)
                return null;

            int index = Convert.ToInt32(par1 ?? 0);
            int count = Convert.ToInt32(par2 ?? str.Length);

            if (index < 0 || index >= str.Length) return "";
            if (count < 0 || index + count > str.Length) count = str.Length - index;

            return str.Substring(index, count);
        }

        public string ParName(int index)
        {
            return index switch
            {
                0 => "string",
                1 => "index",
                2 => "count",
                _ => ""
            };
        }
    }
}
